oap-server/server-library/library-elasticsearch-client/src/main/java/org/apache/skywalking/library/elasticsearch/bulk/BulkProcessor.java

```diff
@@ -94,14 +100,16 @@ public CompletableFuture<Void> add(UpdateRequest request) {
     private CompletableFuture<Void> internalAdd(Object request) {
         requireNonNull(request, "request");
         final CompletableFuture<Void> f = new CompletableFuture<>();
+        int len = toByteArray(request).length;
```

Comment:
@kezhenxu94 I have a little concern about the payload of this, which pushed a request to byte[] but only for length.

Reply:
I am feeling we want this feature, we should somehow work with ElasticSearch client, rather than run serialization twice for a bulk.

Reply:
In another word, we should make `sizeOfBytes` a protection mechanism rather than a flush mechanism. Add this into `BulkProcessor#doFlush` to control the final size of the HTTP request.

@chenhaipeng What do you think?

Reply:
i think so , where is it better to control sizeOfBytes？

Reply:
> i think so , where is it better to control sizeOfBytes？

If a protection mechanism is added when sending, then an unpacking may be required, maybe which increase the complexity

Reply:
I think `BulkProcessor#doFlush` includes all logic you need. You should have everything you need for. I don't think you need to unpack from there. You only have to do is adding the mechanism to build one request(current) or more.

From my understanding, this mechanism is to resolve too large an HTTP body issue, right?

Reply:
Then, changing `BulkProcessor#doFlush` should be good enough and easy enough. Could you try that?

Reply:
I use elasticsearch in the production environment，I solved it simply as above，but this problem may also occur with other storage type

Reply:
> Then, changing `BulkProcessor#doFlush` should be good enough and easy enough. Could you try that?

ok ,i try it

Reply:
> I use elasticsearch in the production environment，I solved it simply as above，but this problem may also occur with other storage type

That is fine. We don't need to resolve an issue that doesn't happen(at least in some known scenarios). Let's focus on fixing this on ElasticSearch.

Reply:
> I use elasticsearch in the production environment，I solved it simply as above，but this problem may also occur with other storage type

i fix it ,please check again

Reply:
> yes , this mechanism is to resolve too large an HTTP body

As you set the default value to 5M, could you share what is the size limitation?

Reply:
> > yes , this mechanism is to resolve too large an HTTP body
> 
> As you set the default value to 5M, could you share what is the size limitation?

there is elastic http limit ,Defaults to 100mb, I think it should not exceed 10m. Kafka has a similar limit of 10m.
https://www.elastic.co/guide/en/elasticsearch/reference/7.17/modules-network.html#http-settings

Reply:
So, you are facing a 100m limitation? I am feeling 5m is very small, and if there are many tags/logs in the segment/log, we may flush a few data into the storage.

Reply:
In a production environment, 5m is about 3000 per dimension. I think in extreme cases, with too many tags, it might cause oversized. It is reasonable to set 10m, https://cloud.tencent.com/document/product/845/56274

Reply:
>  5m is about 3000 per dimension

What does this mean?

Reply:
I set the request body to 5mB, about 3000 docs on average

Reply:
OK, 3k-10k docs should not be an issue. After all, the bulk mixed metrics, traces and logs. Let's keep the default size of 10m. For today's network performance, a 10m flush should not be a problem.